File: apps/react-native/src/bridge/NativeFRWBridge.ts
import type {
  NewKeyInfo as SharedNewKeyInfo,
  RecentContactsResponse,
  AccountKeySignature as SharedAccountKeySignature,
  Currency as SharedCurrency,
  EnvironmentVariables as SharedEnvironmentVariables,
  WalletAccount,
  WalletAccountsResponse,
  WalletProfilesResponse,
  CreateAccountResponse as SharedCreateAccountResponse,
  SeedPhraseGenerationResponse as SharedSeedPhraseGenerationResponse,
  MigrationAssetsData as SharedMigrationAssetsData,
} from '@onflow/frw-types';
import { type NativeScreenName as SharedNativeScreenName } from '@onflow/frw-types';
import type { TurboModule } from 'react-native';
import { TurboModuleRegistry } from 'react-native';

/**
 * Local interfaces for Codegen - must be defined in the same file
 * @see {@link SharedEnvironmentVariables}, {@link SharedCurrency}, {@link SharedCreateAccountResponse}, {@link SharedSeedPhraseGenerationResponse}, and {@link SharedNativeScreenName} in @onflow/frw-types
 *
 * React Native Codegen limitation: Cannot resolve imported types.
 * These must stay in sync with the source types manually.
 */
interface EnvironmentVariables {
  NODE_API_URL: string;
  GO_API_URL: string;
  INSTABUG_TOKEN: string;
  MIXPANEL_TOKEN?: string;
  CADENCE_INBOX?: boolean;
}

interface Currency {
  name: string;
  symbol: string;
  rate: string;
}

interface AccountKey {
  publicKey: string;
  signAlgo?: number;
  hashAlgo?: number;
  weight?: number;
}

interface AccountKeySignature {
  public_key: string;
  hash_algo: number;
  sign_algo: number;
  signature: string;
  sign_message?: string;
  weight?: number;
}

interface NewKeyInfo {
  seedphrase: string;
  flowKey: AccountKey;
}

/**
 * Response from getV4RegistrationSignatures containing all signatures needed for v4 API
 */
interface V4RegistrationSignaturesResponse {
  flowSignature: string;
  evmSignature: string;
  eoaAddress: string;
}

interface CreateAccountResponse {
  success: boolean;
  address: string | null;
  username: string | null;
  accountType: 'full' | 'hardware' | null; // full = mnemonic-based account, hardware = secure enclave
  txId: string | null;
  error: string | null;
}

interface SeedPhraseGenerationResponse {
  mnemonic: string;
  accountKey: AccountKey;
  drivepath: string;
  /** Pre-derived EVM/EOA address from BIP44 path m/44'/60'/0'/0/0 for faster display */
  evmAddress?: string;
}

interface MigrationAssetsData {
  erc20: Array<{ address: string; amount: string }>;
  erc721: Array<{ address: string; id: string }>;
  erc1155: Array<{ address: string; id: string; amount: string }>;
}

/**
 * Local type for NativeScreenName - must match enum values from @onflow/frw-types
 * @see {@link SharedNativeScreenName}
 */
type NativeScreenName =
  | 'multiBackup'
  | 'deviceBackup'
  | 'seedPhraseBackup'
  | 'backupOptions'
  | 'walletRestore'
  | 'recoveryPhraseRestore'
  | 'keyStoreRestore'
  | 'privateKeyRestore'
  | 'googleDriveRestore'
  | 'icloudRestore'
  | 'multiRestore';

// Compile-time sync validation
const _syncCheck: EnvironmentVariables = {} as SharedEnvironmentVariables;
const _reverseSyncCheck: SharedEnvironmentVariables = {} as EnvironmentVariables;
const _currencySyncCheck: Currency = {} as SharedCurrency;
const _currencyReverseSyncCheck: SharedCurrency = {} as Currency;
const _newKeyInfoSyncCheck: NewKeyInfo = {} as SharedNewKeyInfo;
const _newKeyInfoReverseSyncCheck: SharedNewKeyInfo = {} as NewKeyInfo;
const _accountKeySignatureSyncCheck: AccountKeySignature = {} as SharedAccountKeySignature;
const _accountKeySignatureReverseSyncCheck: SharedAccountKeySignature = {} as AccountKeySignature;
const _createAccountSyncCheck: CreateAccountResponse = {} as SharedCreateAccountResponse;
const _createAccountReverseSyncCheck: SharedCreateAccountResponse = {} as CreateAccountResponse;
const _seedPhraseSyncCheck: SeedPhraseGenerationResponse = {} as SharedSeedPhraseGenerationResponse;
const _migrationAssetsSyncCheck: MigrationAssetsData = {} as SharedMigrationAssetsData;
const _migrationAssetsReverseSyncCheck: SharedMigrationAssetsData = {} as MigrationAssetsData;

// NativeScreenName validation - ensures local union matches SharedNativeScreenName enum values
type SharedNativeScreenNameValues = `${SharedNativeScreenName}`;
const _nativeScreenNameSyncCheck: NativeScreenName = {} as SharedNativeScreenNameValues;
const _nativeScreenNameReverseSyncCheck: SharedNativeScreenNameValues = {} as NativeScreenName;

export interface Spec extends TurboModule {
  getSelectedAddress(): string | null;
  getDebugAddress(): string | null;
  getNetwork(): string;
  getJWT(): Promise<string>;
  getVersion(): string;
  getBuildNumber(): string;
  getLanguage(): string;
  // Turbo Modules do not support Uint8Array or ArrayBuffer, so we need to convert to hex string instead
  sign(hexData: string): Promise<string>;
  ethSign(hexData: string, address: string | null): Promise<string>;
  getRecentContacts(): Promise<RecentContactsResponse>;
  // Wallet accounts method
  getWalletAccounts(): Promise<WalletAccountsResponse>;
  getSignKeyIndex(): number;
  // QR code scanning method
  scanQRCode(): Promise<string>;
  // Close react native method
  closeRN(id?: string | null): void;

  // Update dialog action click callback
  onUpdateDialogActionPress(
    actionType: 'external' | 'internal' | 'deeplink',
    actionUrl?: string | null,
    actionText?: string | null
  ): void;

  // Close react native and enable NFT collection storage
  closeRNWithNFT(id?: string | null): void;

  // Free gas settings method
  isFreeGasEnabled(): Promise<boolean>;
  // Listen to a transaction
  listenTransaction(txid: string): void;
  // Get environment variables
  getEnv(): EnvironmentVariables;
  // Get selected account
  getSelectedAccount(): Promise<WalletAccount>;
  getMigrationAssets(sourceAddress: string): Promise<MigrationAssetsData>;
  refreshCoaAfterMigration(): Promise<void>;
  getCurrency(): Currency;
  getTokenRate(token: string): string;
  getWalletProfiles(): Promise<WalletProfilesResponse>;
  // Profile management - recovery flow
  getRecoverableProfiles(): Promise<WalletProfilesResponse>;
  switchToProfile(userId: string): Promise<void>;
  // Device info method
  getDeviceId(): string;
  // Toast methods
  showToast(
    title: string,
    message?: string | null,
    type?: 'success' | 'error' | 'warning' | 'info',
    duration?: number | null
  ): void;
  hideToast(id: string): void;
  clearAllToasts(): void;
  // Native logging method for additional platform-specific logging
  logToNative(
    level: 'debug' | 'info' | 'warn' | 'error',
    message: string,
    args: ReadonlyArray<string>
  ): void;

  // Key rotation methods
  createSeedKey(strength: number): Promise<NewKeyInfo>;
  saveNewKey(key: NewKeyInfo): Promise<void>;
  removeOldKey(address: string, publicKey: string): Promise<void>;
  signRotationRequest(address: string, signatureData: string): Promise<AccountKeySignature>;

  // Onboarding methods
  registerSecureTypeAccount(username: string): Promise<CreateAccountResponse>;
  initSecureEnclaveWallet(
    txId: string
  ): Promise<{ success: boolean; address: string | null; error: string | null }>;
  generateSeedPhrase(strength?: number | null): Promise<SeedPhraseGenerationResponse>;
  /**
   * Get all signatures needed for v4 API registration
   * Signs in anonymously to Firebase, gets JWT, and signs it with both Flow and EVM keys derived from mnemonic
   * @param mnemonic - The recovery phrase to derive signing keys from
   * @returns Promise with flowSignature, evmSignature, and eoaAddress
   */
  getV4RegistrationSignatures(mnemonic: string): Promise<V4RegistrationSignaturesResponse>;
  signInWithCustomToken(customToken: string): Promise<void>;
  saveMnemonic(
    mnemonic: string,
    customToken: string,
    txId: string,
    username: string,
    evmAddress?: string | null
  ): Promise<void>;
  requestNotificationPermission(): Promise<boolean>;
  checkNotificationPermission(): Promise<boolean>;
  // Keystore migration
  keystoreMigration?(): Promise<void>;

  // Screen security
  setScreenSecurityLevel(level: 'normal' | 'secure'): void;
  // Launch native screen method - uses local NativeScreenName type for Codegen compatibility
  launchNativeScreen(screenName: NativeScreenName, params?: string | null): void;
}

export default TurboModuleRegistry.getEnforcing<Spec>('NativeFRWBridge');

File: app/src/main/java/com/flowfoundation/wallet/reactnative/bridge/NativeFRWBridge.kt
package com.flowfoundation.wallet.reactnative.bridge

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.flowfoundation.wallet.reactnative.bridge.handlers.AccountBridgeHandler
import com.flowfoundation.wallet.reactnative.bridge.handlers.AuthBridgeHandler
import com.flowfoundation.wallet.reactnative.bridge.handlers.UIBridgeHandler
import com.flowfoundation.wallet.reactnative.bridge.handlers.UtilsBridgeHandler
import com.flowfoundation.wallet.reactnative.bridge.handlers.WalletBridgeHandler
import com.flowfoundation.wallet.utils.logd
import com.google.gson.Gson
import org.json.JSONArray
import org.json.JSONObject

class NativeFRWBridge(reactContext: ReactApplicationContext) : NativeFRWBridgeSpec(reactContext) {

    private val TAG = "NativeFRWBridge"

    // Delegate handlers for different bridge functionality domains
    private val utilsHandler = UtilsBridgeHandler(reactContext)
    private val accountHandler = AccountBridgeHandler(reactContext)
    private val authHandler = AuthBridgeHandler(reactContext)
    private val uiHandler = UIBridgeHandler(reactContext)
    private val walletHandler = WalletBridgeHandler(reactContext)

    /**
     * Send an event to React Native JavaScript
     * @param eventName The name of the event
     * @param params The parameters to send with the event
     */
    fun sendEvent(eventName: String, params: WritableMap?) {
        try {
            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit(eventName, params)
        } catch (e: Exception) {
            logd(TAG, "Failed to send event $eventName: ${e.message}")
        }
    }

    init {
        logd(TAG, "NativeFRWBridge initialized with context: ${reactContext != null}")
        logd(TAG, "React context is active: ${reactContext.hasActiveCatalystInstance()}")
    }

    override fun getName(): String = utilsHandler.getName()

    override fun getSelectedAddress(): String? = accountHandler.getSelectedAddress()

    override fun getDebugAddress(): String? = accountHandler.getDebugAddress()

    override fun getNetwork(): String = utilsHandler.getNetwork()

    override fun getJWT(promise: Promise) = authHandler.getJWT(promise)

    override fun getVersion(): String = utilsHandler.getVersion()

    override fun getBuildNumber(): String = utilsHandler.getBuildNumber()

    override fun getLanguage(): String? = utilsHandler.getLanguage()

    override fun getDeviceId(): String = utilsHandler.getDeviceId()

    override fun sign(hexData: String, promise: Promise) = walletHandler.sign(hexData, promise)

    override fun ethSign(hexData: String?, promise: Promise?) = walletHandler.ethSign(hexData, promise)

    override fun listenTransaction(txid: String) = walletHandler.listenTransaction(txid)

    override fun scanQRCode(promise: Promise) = uiHandler.scanQRCode(promise)

    override fun getRecentContacts(promise: Promise) = accountHandler.getRecentContacts(promise, ::bridgeModelToWritableMap)

    override fun getWalletAccounts(promise: Promise) = accountHandler.getWalletAccounts(promise, ::bridgeModelToWritableMap)

    override fun closeRN(id: String?) = uiHandler.closeRN(id)

    override fun getSignKeyIndex(): Double = accountHandler.getSignKeyIndex()

    override fun isFreeGasEnabled(promise: Promise) = utilsHandler.isFreeGasEnabled(promise)

    override fun getEnv(): WritableMap = utilsHandler.getEnv(::bridgeModelToWritableMap)

    override fun getSelectedAccount(promise: Promise) = accountHandler.getSelectedAccount(promise, ::bridgeModelToWritableMap)

    override fun getCurrency(): WritableMap = utilsHandler.getCurrency(::bridgeModelToWritableMap)

    override fun getTokenRate(token: String): String = utilsHandler.getTokenRate(token)

    private val gson = Gson()

    // Helper method to convert Bridge models to React Native data
    private fun bridgeModelToWritableMap(model: Any): WritableNativeMap {
        return try {
            val json = gson.toJson(model)
            jsonToWritableMap(json)
        } catch (e: Exception) {
            println("Error converting bridge model to WritableMap: ${e.message}")
            e.printStackTrace()
            WritableNativeMap()
        }
    }

    // Helper method to convert JSON string to WritableMap
    private fun jsonToWritableMap(jsonString: String): WritableNativeMap {
        val map = WritableNativeMap()
        try {
            val jsonObject = JSONObject(jsonString)
            jsonObject.keys().forEach { key ->
                val value = jsonObject.get(key)
                when (value) {
                    is String -> map.putString(key, value)
                    is Boolean -> map.putBoolean(key, value)
                    is Int -> map.putInt(key, value)
                    is Long -> map.putDouble(key, value.toDouble())
                    is Float -> map.putDouble(key, value.toDouble())
                    is Double -> map.putDouble(key, value)
                    is JSONArray -> map.putArray(key, jsonArrayToWritableArray(value))
                    is JSONObject -> map.putMap(key, jsonToWritableMap(value.toString()))
                    JSONObject.NULL -> map.putNull(key)
                    null -> map.putNull(key)
                    else -> {
                        // Handle any other types by converting to string
                        map.putString(key, value.toString())
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return map
    }

    // Helper method to convert JSONArray to WritableArray
    private fun jsonArrayToWritableArray(jsonArray: JSONArray): WritableNativeArray {
        val array = WritableNativeArray()
        try {
            for (i in 0 until jsonArray.length()) {
                val value = jsonArray.get(i)
                when (value) {
                    is String -> array.pushString(value)
                    is Boolean -> array.pushBoolean(value)
                    is Int -> array.pushInt(value)
                    is Long -> array.pushDouble(value.toDouble())
                    is Float -> array.pushDouble(value.toDouble())
                    is Double -> array.pushDouble(value)
                    is JSONObject -> array.pushMap(jsonToWritableMap(value.toString()))
                    is JSONArray -> array.pushArray(jsonArrayToWritableArray(value))
                    JSONObject.NULL -> array.pushNull()
                    null -> array.pushNull()
                    else -> {
                        // Handle any other types by converting to string
                        array.pushString(value.toString())
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return array
    }


    override fun getWalletProfiles(promise: Promise) = accountHandler.getWalletProfiles(promise, ::bridgeModelToWritableMap)

    override fun getRecoverableProfiles(promise: Promise) = accountHandler.getRecoverableProfiles(promise, ::bridgeModelToWritableMap)

    override fun switchToProfile(userId: String, promise: Promise) = accountHandler.switchToProfile(userId, promise)

    override fun showToast(title: String, message: String?, type: String?, duration: Double?) = uiHandler.showToast(title, message, type, duration)

    override fun hideToast(id: String) = uiHandler.hideToast(id)

    override fun clearAllToasts() = uiHandler.clearAllToasts()

    override fun registerSecureTypeAccount(username: String, promise: Promise) = authHandler.registerSecureTypeAccount(username, promise, ::sendEvent)

    override fun initSecureEnclaveWallet(txId: String, promise: Promise) = authHandler.initSecureEnclaveWallet(txId, promise)

    override fun generateSeedPhrase(strength: Double?, promise: Promise) = authHandler.generateSeedPhrase(strength, promise, ::bridgeModelToWritableMap)

    // TODO: Add getRegistrationSignature to TypeScript spec and regenerate codegen to make this a bridge method
    fun getRegistrationSignature(mnemonic: String, promise: Promise) = authHandler.getRegistrationSignature(mnemonic, promise)

    override fun signInWithCustomToken(customToken: String, promise: Promise) = authHandler.signInWithCustomToken(customToken, promise)

    override fun saveMnemonic(mnemonic: String, customToken: String, txId: String, username: String, promise: Promise) = authHandler.saveMnemonic(mnemonic, customToken, txId, username, promise, ::sendEvent)

    override fun requestNotificationPermission(promise: Promise) = utilsHandler.requestNotificationPermission(promise)

    override fun checkNotificationPermission(promise: Promise) = utilsHandler.checkNotificationPermission(promise)

    override fun logToNative(level: String, message: String, args: ReadableArray) = utilsHandler.logToNative(level, message, args)

    override fun setScreenSecurityLevel(level: String) = utilsHandler.setScreenSecurityLevel(level)

    override fun launchNativeScreen(screenName: String, params: String?) = uiHandler.launchNativeScreen(screenName, params)

    companion object {
        const val NAME = "NativeFRWBridge"
    }
}
